File: mesh2scattering/Output2HRTF/output2hrtf.m
function output2hrtf(folder)
%   [] = OUTPUT2HRTF_MAIN(folder)
%
%   Process NumCalc output and write data to disk. All parameters are read
%   from parameters.json.
%
%   Input:
%       folder ... The Mesh2HRTF project folder [string, the default is
%                  the current working directory]

%% ----------------------------load meta data------------------------------

if ~exist('folder', 'var')
    folder = pwd;
end

% load parameters
if ~isfile(fullfile(folder, 'parameters.json'))
    error('parameters.json not found.')
end
params = fullfile(folder, 'parameters.json');
fid = fopen(params);
params = fread(fid,inf);
params = char(params');
fclose(fid);
params = jsondecode(params);
if all(size(params.sourceCenter) == [3 1])
    params.sourceCenter = params.sourceCenter';
end

% output directory
if ~exist(fullfile(folder, 'Output2HRTF'), 'dir')
    mkdir(fullfile(folder, 'Output2HRTF'))
end

fprintf('Load meta data ...\n');

% get the evaluation grids
fprintf('Get evaluation grids ...\n');
evaluationGrids = dir(fullfile(folder, 'EvaluationGrids'));
evaluationGrids = evaluationGrids(~cellfun(@(x) strncmp(x, '.', 1), {evaluationGrids.name}));

evaluationGridsNumNodes = 0;
for ii=1:length(evaluationGrids)
    tmpNodes=importdata(fullfile(folder, 'EvaluationGrids', evaluationGrids(ii).name, 'Nodes.txt'),' ',1);
    tmpElements=importdata(fullfile(folder, 'EvaluationGrids', evaluationGrids(ii).name, 'Elements.txt'),' ',1);
    evaluationGrids(ii).nodes = tmpNodes.data;
    evaluationGrids(ii).elements = tmpElements.data;
    evaluationGrids(ii).num_nodes = size(tmpNodes.data, 1);
    evaluationGridsNumNodes = evaluationGridsNumNodes + evaluationGrids(ii).num_nodes;
end

% get the object mesh
fprintf('Get object meshes ...\n');
objectMeshes = dir(fullfile(folder, 'ObjectMeshes'));
objectMeshes = objectMeshes(~cellfun(@(x) strncmp(x, '.', 1), {objectMeshes.name}));

objectMeshesNumNodes = 0;
for ii=1:length(objectMeshes)
    tmpNodes=importdata(fullfile(folder, 'ObjectMeshes', objectMeshes(ii).name, 'Nodes.txt'),' ',1);
    tmpElements=importdata(fullfile(folder, 'ObjectMeshes', objectMeshes(ii).name, 'Elements.txt'),' ',1);
    objectMeshes(ii).nodes = tmpNodes.data;
    objectMeshes(ii).elements = tmpElements.data;
    objectMeshes(ii).num_nodes = size(tmpNodes.data, 1);
    objectMeshesNumNodes = objectMeshesNumNodes + objectMeshes(ii).num_nodes;
end

% build frequency vector
frequencies = params.frequencies;

clear ii tmpNodes tmpElements

%% Read computational effort
fprintf('Loading computational effort data ...\n');
% check for folder be.out and read computation time from every NC*.out file
for ii = 1:params.numSources
    computationTime{ii} = [];
    boundaryElements = dir([folder, filesep, 'NumCalc', filesep, 'source_', num2str(ii), filesep, 'NC*.out']);

    NC_all_flag = 0; NC_all_idx = []; NC_all_date = [];
    NC_single_flag = 0; NC_single_idx = []; NC_single_date = [];

    % read NC*.out configurations
    for jj = 1:size(boundaryElements, 1)
        if ~isempty(regexp(boundaryElements(jj).name, '^(NC.out)$')) % NC.out exists
            NC_all_flag = 1;
            NC_all_idx = jj;
            NC_all_date = datenum(boundaryElements(jj).date);
        elseif ~isempty(regexp(boundaryElements(jj).name, '(NC.)\S+(.out)')) % NC*.out exists
            NC_single_flag = 1;
            NC_single_idx = [NC_single_idx, jj];
            NC_single_date = [NC_single_date, datenum(boundaryElements(jj).date)];
        end
    end
    clear jj

    fprintf(['Reading computation time from source ', num2str(ii), '\n']);
    % read from .out files according to latest calculation
    if NC_all_flag && ~NC_single_flag % only NC.out exists
        % read computation time
        tmp=write_output_report([folder, filesep, 'NumCalc', filesep, 'source_', ...
            num2str(ii), filesep, boundaryElements(NC_all_idx).name]);
        computationTime{ii}=[computationTime{ii}; tmp];
    elseif NC_single_flag && ~NC_all_flag % only NC*.out exist
        % read single files
        for jj = 1:length(NC_single_idx)
            % read computation time
            tmp=write_output_report([folder, filesep, 'NumCalc', filesep, 'source_', ...
                num2str(ii), filesep, boundaryElements(NC_single_idx(jj)).name]);
            computationTime{ii}=[computationTime{ii}; tmp];
        end
    elseif (NC_all_flag && NC_single_flag) && any(NC_all_date < NC_single_date) % both exist, at least one NC*.out exists whose frequencies replace the ones in NC.out
        % start with NC.out ...
        tmp=write_output_report([folder, filesep, 'NumCalc', filesep, 'source_', ...
            num2str(ii), filesep, boundaryElements(NC_all_idx).name]);
        computationTime{ii}=[computationTime{ii}; tmp];
        % ... then read single files that are newer than NC.out
        for jj = 1:length(NC_single_idx)
            if NC_all_date < NC_single_date(jj)
                % read computation time
                tmp=write_output_report([folder, filesep, 'NumCalc', filesep, 'source_', ...
                    num2str(ii), filesep, boundaryElements(NC_single_idx(jj)).name]);
                computationTime{ii}(tmp(:,1),:) = tmp; % replace data only for current frequencies
            end
        end
    elseif (NC_all_flag && NC_single_flag) && all(NC_all_date > NC_single_date) % both exist, NC.out newest file
        % same as only NC.out exists
        % read computation time
        tmp=write_output_report([folder, filesep, 'NumCalc', filesep, 'source_', ...
            num2str(ii), filesep, boundaryElements(NC_all_idx).name]);
        computationTime{ii}=[computationTime{ii}; tmp];
    else % what possible case is this?
        error('This case is not yet implemented. Please open an issue at the project page: https://github.com/Any2HRTF/Mesh2HRTF/issues');
    end
    clear jj

    % check for error potential in numerical calculation
    warning('off', 'backtrace');
    iter_error_idx = find(computationTime{ii}(:,9));
    rel_error_idx = find(computationTime{ii}(:,7) > 1e-9);
    if ~isempty(iter_error_idx)
        for jj=1:length(iter_error_idx)
            warning(['Number of iterations for frequency ', num2str(computationTime{ii}(iter_error_idx(jj), 2)), ...
                ' Hz has reached the maximum.'], '\n');
        end
        clear jj
    end
    if ~isempty(rel_error_idx)
        for jj=1:length(rel_error_idx)
            warning(['Relative error for frequency ', num2str(computationTime{ii}(rel_error_idx(jj), 2)), ...
                ' Hz is greater than 1e-9.'], '\n');
        end
    end
    warning('on', 'backtrace');
    % sort computation time after frequency steps
    [~, sort_idx] = sort(computationTime{ii}(:,1));
    computationTime{ii} = computationTime{ii}(sort_idx,:);
end

fprintf('Write computation time to .mat file ...\n');
description={'Frequency index','Frequency','Building','Solving','Postprocessing','Total', 'relative error', 'number of iterations', 'maximum number of iterations reached'};
save(fullfile(folder, 'Output2HRTF', 'computationTime.mat'), 'description', 'computationTime', '-v6');
clear ii jj description computationTime tmp iter_error_idx rel_error_idx

%% Load ObjectMesh data
fprintf('Loading ObjectMesh data for ');
for ii=1:params.numSources
    fprintf(['source ', num2str(ii), '\n']);
    pressure{ii}=Output2HRTF_Load([folder, filesep, 'NumCalc', filesep, 'source_', num2str(ii), ...
        filesep, 'be.out', filesep], 'pBoundary', params.numFrequencies);
    clear data
end

clear ii

fprintf('Save ObjectMesh data ...\n');
cnt = 0;
for ii = 1:numel(objectMeshes)
    nodes=objectMeshes(ii).nodes;
    elements=objectMeshes(ii).elements;
    element_data=pressure;
    for jj = 1:numel(pressure)
        element_data{jj} = element_data{jj}(:, cnt+1:cnt+size(elements,1));
    end
    save(fullfile(folder, 'Output2HRTF', ['ObjectMesh_' objectMeshes(ii).name '.mat']), ...
        'nodes','elements','frequencies','element_data');
    cnt = cnt + size(elements,1);
end

clear pressure nodes elements ii jj cnt idx element_data

%% Load EvaluationGrid data
if ~isempty(evaluationGrids)
    fprintf('Loading evaluation grids for data in ');
    for ii=1:params.numSources
        fprintf(['source ', num2str(ii), '\n']);
        pressure(:,:,ii)=Output2HRTF_Load([folder, filesep, 'NumCalc', filesep, 'source_', num2str(ii), ...
            filesep, 'be.out', filesep], 'pEvalGrid', params.numFrequencies);
    end
end
clear ii

% save to struct
fprintf('Save EvaluationGrid data ...\n');
cnt = 0;
for ii = 1:numel(evaluationGrids)
    evaluationGrids(ii).pressure = pressure(:, cnt+1:cnt+evaluationGrids(ii).num_nodes, :);
    cnt = cnt + evaluationGrids(ii).num_nodes;
end


clear ii pressure cnt idx


% reference to pressure in the middle of the head with the head absent
% according to the HRTF definition.
if params.reference
    fprintf('Divide pressure by reference according to HRTF definition ...\n');
    % this might be a parameter in the function call
    refMode = 1;    % 1: reference to only one radius (the smallest found)
    % 2: reference to all indivudal radii

    for ii = 1:numel(evaluationGrids)

        xyz = evaluationGrids(ii).nodes;
        pressure = evaluationGrids(ii).pressure;
        freqMatrix = repmat(frequencies, [1 size(pressure,2) size(pressure,3)]);

        % distance of source positions from the origin
        if refMode == 1
            r = min(sqrt(xyz(:,2).^2 + xyz(:,3).^2 + xyz(:,4).^2));
            r = repmat(r, [size(pressure,1) size(pressure, 2) size(pressure, 3)]);
        else
            r = sqrt(xyz(:,2).^2 + xyz(:,3).^2 + xyz(:,4).^2);
            r = repmat(r', [size(pressure,1) 1 size(pressure, 3)]);
        end

        if strcmp(params.sourceType, 'Left ear') || ...
                strcmp(params.sourceType, 'Right ear') || ...
                strcmp(params.sourceType, 'Both ears')

            volumeFlow = .1 * ones(size(pressure));
            if isfield(params, 'sourceArea')
                for jj = 1:numel(params.sourceArea)
                    volumeFlow(:,:,jj) = volumeFlow(:,:,jj) * params.sourceArea(jj);
                end
            end

            % point source in the origin evaluated at r
            % eq. (6.71) in: Williams, E. G. (1999). Fourier Acoustics.
            ps   = -1j * params.densityOfMedium * 2*pi*freqMatrix .* volumeFlow ./ (4*pi) .* ...
                exp(1j * 2*pi*freqMatrix/params.speedOfSound .* r) ./ r;

        elseif strcmp(params.sourceType, 'Point source')

            amplitude = .1; % hard coded in Mesh2HRTF
            ps = amplitude * exp(1j * 2*pi*freqMatrix/params.speedOfSound .*r) ./ (4 * pi * r);

        elseif strcmp(params.sourceType, 'Plane wave')
            error('Referencing for plane wave source type not yet implemented.\n');
        else
            error('Referencing is currently only implemented for sourceType ''vibratingElement'' and ''pointSource''.\n')
        end

        evaluationGrids(ii).pressure = pressure ./ ps;

    end

    clear r freqMatrix ps ii jj
end

%% Save data as SOFA file
fprintf('Saving complex pressure to SOFA file ...\n')
SOFAstart;

for ii = 1:numel(evaluationGrids)

    xyz = evaluationGrids(ii).nodes;

    % check if .sofa file already exists
    filename = fullfile('Output2HRTF', ['HRTF_' evaluationGrids(ii).name '.sofa']);
    if exist(filename, 'file')
        % if yes, delete and write new object, but ask user
        prompt = [filename, ' already exists. Replace object? [y/n]\n'];
        replace_obj = input(prompt, 's');
        if strcmp(replace_obj, 'y') || strcmp(replace_obj, 'yes')
            delete(filename)
        elseif strcmp(replace_obj, 'n') || strcmp(replace_obj, 'no')
            error('Object was not replaced. Output2HRTF aborted.\n');
        else
            error('Invalid input. Output2HRTF aborted.\n');
        end
    end

    % get SOFA template according to number of sources
    if params.numSources == 2
        Obj = SOFAgetConventions('SimpleFreeFieldHRTF');
    else
        % Save as GeneralTF
        Obj = SOFAgetConventions('GeneralTF');
    end

    % pressure has size N (frequencies) x M (measurements) x R (sources)
    pressure = evaluationGrids(ii).pressure;
    % get dimenson 3 explicitly, because it would be dropped if R=1
    NMR = [size(pressure, 1), size(pressure, 2), size(pressure, 3)];

    % shift dimensions to MRN as expected by SOFA
    % (shifting loses leading and trailing dimensons of size 1)
    pressure = shiftdim(pressure, 1);
    % force dimensions of 1
    pressure = reshape(pressure, NMR(2), NMR(3), NMR(1));

    Obj.GLOBAL_ApplicationName = 'Mesh2HRTF';
    Obj.GLOBAL_ApplicationVersion = params.Mesh2HRTF_Version;
    Obj.GLOBAL_History = 'numerically calculated data';

    Obj.N=frequencies;
    Obj.Data.Real=real(pressure);
    Obj.Data.Imag=imag(pressure);
    Obj.Data.Real_LongName='pressure';
    Obj.Data.Real_Units='pascal';
    Obj.Data.Imag_LongName='pressure';
    Obj.Data.Imag_Units='pascal';

    Obj.ListenerPosition = [0 0 0];
    Obj.SourcePosition=xyz(:,2:4);
    Obj.SourcePosition_Type='cartesian';
    Obj.SourcePosition_Units='metre';
    Obj.ReceiverPosition=params.sourceCenter;
    Obj.ReceiverPosition_Type='cartesian';
    Obj.ReceiverPosition_Units='metre';

    Obj=SOFAupdateDimensions(Obj);
    SOFAsave(fullfile(folder, 'Output2HRTF', ['HRTF_' evaluationGrids(ii).name '.sofa']),Obj);
    fprintf(['HRTF_' evaluationGrids(ii).name '.sofa saved!\n']);
end

clear Obj ii xyz pressure prompt replace_Obj


%% Save time data data as SOFA file
if params.computeHRIRs
    fprintf('Saving HRIR data to SOFA file ...\n')
    for ii = 1:numel(evaluationGrids)
        % check if the frequency vector has the correct format
        if any(abs(diff(frequencies,2)) > .1) || frequencies(1) < .1
            error('The frequency vector must start at a frequency > 0.1 and continue in equidistant steps to the end.\n')
        end

        % check if reference exists
        if ~params.reference
            error('HRIRs can only be computed if refernce=true\n')
        end

        xyz = evaluationGrids(ii).nodes;
        pressure = evaluationGrids(ii).pressure;

        % check if .sofa file already exists
        filename = fullfile('Output2HRTF', ['HRIR_' evaluationGrids(ii).name '.sofa']);
        if exist(filename, 'file')
            % if yes, delete and write new object, but ask user
            prompt = [filename, ' already exists. Replace object? [y/n]\n'];
            replace_obj = input(prompt, 's');
            if strcmp(replace_obj, 'y') || strcmp(replace_obj, 'yes')
                delete(filename)
            elseif strcmp(replace_obj, 'n') || strcmp(replace_obj, 'no')
                error('Object was not replaced. Output2HRTF aborted.\n');
            else
                error('Invalid input. Output2HRTF aborted.\n');
            end
        end
        clear prompt

%         prompt = ['The default sampling frequency is twice the highest occuring frequency. ', ...
%             'In this case fs = ', num2str(2*frequencies(end)), '. \n', ...
%             'Do you want to specify a different sampling frequency? [y/n]\n'];
%         replace_fs = input(prompt, 's');
%         if strcmp(replace_fs, 'y') || strcmp(replace_fs, 'yes')
%             clear prompt
%             prompt = 'Please specify the sampling frequency in Hz: ';
%             fs = input(prompt);
%             if fs < 0
%                 error('The sampling frequency has to be positive.\n');
%             elseif fs < frequencies(end)
%                 warning('The sampling frequency is lower than the highest frequency in the signal. Further processing will potentially lead to aliasing.');
%             end
%             clear prompt
%         elseif strcmp(replace_fs, 'n') || strcmp(replace_fs, 'no')
%             fs = 2*frequencies(end);
%         else
%             error('Invalid input. Writing HRIR to SOFA object aborted.\n');
%         end

        fs = 2*frequencies(end);

        % add 0 Hz bin
        pressure = [ones(1, size(pressure,2), size(pressure,3));
            pressure];
        % make fs/2 real
        pressure(end,:,:) = abs(pressure(end,:,:));
        % mirror the spectrum
        pressure = [pressure; flipud(conj(pressure(2:end-1,:,:)))];
        % ifft (take complex conjugate because sign conventions differ)
        hrir = ifft(conj(pressure), 'symmetric');

        % shift 30 cm to make causal
        % (path differences between the origin and the ear are usually
        % smaller than 30 cm but numerical HRIRs show stringer pre-ringing)
        n_shift = round(.30 / (1/fs * params.speedOfSound));
        hrir = circshift(hrir, n_shift);

        % hrir has size N (samples) x M (measurements) x R (sources)
        % get dimenson 3 explicitly, because it would be dropped if R=1
        NMR = [size(hrir, 1), size(hrir, 2), size(hrir, 3)];

        % shift dimensions to MRN as expected by SOFA
        % (shifting loses leading and trailing dimensons of size 1)
        hrir = shiftdim(hrir, 1);
        % force dimensions of 1
        hrir = reshape(hrir, NMR(2), NMR(3), NMR(1));

        % get SOFA template according to number of sources
        if params.numSources == 2
            Obj = SOFAgetConventions('SimpleFreeFieldHRIR');
        else
            % Save as GeneralFIR
            Obj = SOFAgetConventions('GeneralFIR');
        end

        Obj.GLOBAL_ApplicationName = 'Mesh2HRTF';
        Obj.GLOBAL_ApplicationVersion = params.Mesh2HRTF_Version;
        Obj.GLOBAL_History = 'numerically calculated data';

        Obj.Data.IR=hrir;
        Obj.Data.SamplingRate=fs;
        Obj.Data.Delay=zeros(1, size(hrir, 2));

        Obj.ListenerPosition = [0 0 0];
        Obj.SourcePosition=xyz(:,2:4);
        Obj.SourcePosition_Type='cartesian';
        Obj.SourcePosition_Units='metre';
        Obj.ReceiverPosition=params.sourceCenter;
        Obj.ReceiverPosition_Type='cartesian';
        Obj.ReceiverPosition_Units='metre';

        Obj=SOFAupdateDimensions(Obj);
        SOFAsave(fullfile(folder, 'Output2HRTF', ['HRIR_' evaluationGrids(ii).name '.sofa']),Obj);
        fprintf(['HRIR_' evaluationGrids(ii).name '.sofa saved!\n']);
    end
end

clear Obj ii xyz pressure prompt replace_fs

fprintf('Done\n')

end % end of main function ------------------------------------------------


function data = Output2HRTF_Load(foldername,filename, numFrequencies)
%   data = OUTPUT2HRTF_LOAD(foldername,filename)
%
%   Loads results of the BEM-HRTF calculation from the NumCalc folder.
%
%   Input:
%       foldername ... path to NumCalc output
%       filename ..... either 'pBoundary' or 'pEvalGrid'
%
%   Output:
%       data ......... Matrix of complex values
%           dim1: frequencies
%           dim2: datapoints (complex pressure values)

%% --------------------check number of header lines------------------------
for ii=1:1000
    temp1=importdata([foldername, 'be.1', filesep, filename], ' ', ii);
    if ~isempty(temp1)
        if size(temp1.data,2)==3
            if temp1.data(2,1)-temp1.data(1,1)==1
                numHeaderlines_BE=ii;
                break
            end
        end
    end
end

%% -----------------------------load data----------------------------------
temp1 = importdata([foldername, 'be.1', filesep, filename], ' ', numHeaderlines_BE);

data=zeros(numFrequencies,size(temp1.data,1));

for ii=1:numFrequencies
    tmpData = importdata([foldername, 'be.', num2str(ii), filesep, filename], ' ', numHeaderlines_BE);
    if ~isempty(tmpData)
        data(ii,:)=transpose(tmpData.data(:,2)+1i*tmpData.data(:,3));
    else
        data(ii,:)      = NaN;
    end
end

end %of function

File: mesh2scattering/Output2HRTF/output2hrtf.py
"""
Python tools for Mesh2HRTF including functions to generate SOFA files
containing the HRTF/HRIR data, merge SOFA files containing data for the left
and right ear and generate evaluation grids.
"""
import os
import csv
import warnings
import numpy as np
import json
import sofar as sf
import mesh2scattering as m2s


def output2hrtf(folder=None):
    """
    Process NumCalc output and write data to disk.

    Processing the data is done in the following steps

    1. Read project parameter `from parameters.json`
    2. use :py:func:`~write_output_report` to parse files in
       project_folder/NumCalc/source_*/NC*.out, write project report to
       project_folder/Output2HRTF/report_source_*.csv. Raise a warning if any
       issues were detected and write report_issues.txt to the same folder
    3. Read simulated pressures from project_folder/NumCalc/source_*/be.out.
       This and the following steps are done, even if an issue was detected in
       the previous step
    4. use :py:func:`~mesh2hrtf.reference_hrtfs` and
       :py:func:`~mesh2hrtf.compute_hrirs` to save the results to SOFA files

    Parameters
    ----------
    folder : str, optional
        The path of the Mesh2HRTF project folder, i.e., the folder containing
        the subfolders EvaluationsGrids, NumCalc, and ObjectMeshes. The
        default, ``None`` uses the current working directory.
    """

    # check input
    if folder is None:
        folder = os.getcwd()

    # check and load parameters, required parameters are:
    # Mesh2HRTF_version, reference, computeHRIRs, speedOfSound, densityOfAir,
    # numSources, sourceType, sourceCenter, sourceArea,
    # numFrequencies, frequencies
    params = os.path.join(folder, "parameters.json")
    if not os.path.isfile(params):
        raise ValueError((
            f"The folder {folder} is not a valid Mesh2HRTF project. "
            "It must contain the file 'parameters.json'"))

    with open(params, "r") as file:
        params = json.load(file)

    # output directory
    if not os.path.exists(os.path.join(folder, 'Output2HRTF')):
        os.makedirs(os.path.join(folder, 'Output2HRTF'))

    # write the project report and check for issues
    print('\n Writing the project report ...')
    found_issues, report = m2s.write_output_report(folder)

    if found_issues:
        warnings.warn(report)

    # get the evaluation grids
    print('\n Loading the EvaluationGrids ...')
    evaluationGrids, _ = _read_nodes_and_elements(
        os.path.join(folder, 'EvaluationGrids'))

    # Load EvaluationGrid data
    if not len(evaluationGrids) == 0:
        print('\nLoading data for the evaluation grids ...')

        pressure, _ = _read_numcalc_data(
            params["numSources"], params["numFrequencies"],
            folder, 'pEvalGrid')

    # save to struct
    cnt = 0
    for grid in evaluationGrids:
        evaluationGrids[grid]["pressure"] = pressure[
            cnt:cnt+evaluationGrids[grid]["num_nodes"], :, :]

        cnt = cnt + evaluationGrids[grid]["num_nodes"]

    del grid, pressure, cnt

    # process BEM data for writing HRTFs and HRIRs to SOFA files
    for grid in evaluationGrids:

        print(f'\nSaving HRTFs for EvaluationGrid "{grid}" ...\n')

        # get pressure as SOFA object (all following steps are run on SOFA
        # objects. This way they are available to other users as well)
        sofa = _get_sofa_object(
            evaluationGrids[grid]["pressure"],
            evaluationGrids[grid]["nodes"][:, 1:4], "cartesian",
            params["sourceCenter"], "HRTF", params["Mesh2HRTF_Version"],
            frequencies=params["frequencies"])

        # reference to sound pressure at the center of the head
        if params["reference"]:
            sofa = m2s.reference_hrtfs(
                sofa, params["sourceType"], params["sourceArea"],
                params["speedOfSound"], params["densityOfMedium"], mode="min")

        # write HRTF data to SOFA file
        sf.write_sofa(os.path.join(
            folder, 'Output2HRTF', f'HRTF_{grid}.sofa'), sofa)

        # calculate and write HRIRs
        if params["computeHRIRs"]:
            if not params["reference"]:
                raise ValueError("Computing HRIRs requires prior referencing")

            # calculate shift value (equivalent to a 30 cm shift)
            sampling_rate = round(2*sofa.N[-1])
            n_shift = int(np.round(
                .30 / (1/sampling_rate * params["speedOfSound"])))

            sofa = m2s.compute_hrirs(sofa, n_shift)
            sf.write_sofa(os.path.join(
                folder, 'Output2HRTF', f'HRIR_{grid}.sofa'), sofa)

    print('Done\n')


def _read_nodes_and_elements(folder, objects=None):
    """
    Read the nodes and elements of the evaluation grids or object meshes.

    Parameters
    ----------
    folder : str
        Folder containing the object. Must end with EvaluationGrids or
        Object Meshes
    objects : str, options
        Name of the object. The default ``None`` reads all objects in folder

    Returns
    -------
    grids : dict
        One item per object (with the item name being the object name). Each
        item has the sub-items `nodes`, `elements`, `num_nodes`, `num_elements`
    gridsNumNodes : int
        Number of nodes in all grids
    """
    # check input
    if os.path.basename(folder) not in ['EvaluationGrids', 'ObjectMeshes']:
        raise ValueError('folder must be EvaluationGrids or ObjectMeshes!')

    if objects is None:
        objects = os.listdir(folder)
        # discard hidden folders that might occur on Mac OS
        objects = [o for o in objects if not o.startswith('.')]
    elif isinstance(objects, str):
        objects = [objects]

    grids = {}
    gridsNumNodes = 0

    for grid in objects:
        tmpNodes = np.loadtxt(os.path.join(
            folder, grid, 'Nodes.txt'),
            delimiter=' ', skiprows=1)

        tmpElements = np.loadtxt(os.path.join(
            folder, grid, 'Elements.txt'),
            delimiter=' ', skiprows=1)

        grids[grid] = {
            "nodes": tmpNodes,
            "elements": tmpElements,
            "num_nodes": tmpNodes.shape[0],
            "num_elements": tmpElements.shape[0]}

        gridsNumNodes += grids[grid]['num_nodes']

    return grids, gridsNumNodes


def _read_numcalc_data(numSources, numFrequencies, folder, data):
    """Read the sound pressure on the object meshes or evaluation grid."""
    pressure = []

    if data not in ['pBoundary', 'pEvalGrid', 'vBoundary', 'vEvalGrid']:
        raise ValueError(
            'data must be pBoundary, pEvalGrid, vBoundary, or vEvalGrid')

    print('\n Loading %s data ...' % data)
    for source in range(numSources):
        print('\n    Source %d ...' % (source+1))

        tmpFilename = os.path.join(
            folder, 'NumCalc', f'source_{source+1}', 'be.out')
        tmpPressure, indices = _output_to_hrtf_load(
            tmpFilename, data, numFrequencies)

        pressure.append(tmpPressure)

    pressure = np.transpose(np.array(pressure), (2, 0, 1))

    return pressure, indices


def _get_sofa_object(data, source_position, source_position_type,
                     receiver_position, mode, Mesh2HRTF_version,
                     frequencies=None, sampling_rate=None):
    """
    Write complex pressure or impulse responses to a SOFA object.

    Parameters
    ----------
    data : numpy array
        The data as an array of shape (MRE)
    evaluation_grid : numpy array
        The evaluation grid in Cartesian coordinates as an array of shape (MC)
    receiver_position : numpy array
        The position of the receivers (ears) in Cartesian coordinates
    mode : str
        "HRTF" to save HRTFs, "HRIR" to save HRIRs
    Mesh2HRTF_version : str
    frequencies : numpy array
        The frequencies at which the HRTFs were calculated. Required if mode is
        "HRTF"
    sampling_rate :
        The sampling rate. Required if mode is "HRIR"

    Returns
    -------
    sofa : sofar.Sofa object
    """

    # get source coordinates in spherical convention
    if source_position_type == "cartesian":
        xyz = source_position

        radius = np.sqrt(xyz[:, 0]**2 + xyz[:, 1]**2 + xyz[:, 2]**2)
        z_div_r = np.where(radius != 0, xyz[:, 2] / radius, 0)
        elevation = 90 - np.arccos(z_div_r) / np.pi * 180
        azimuth = np.mod(np.arctan2(xyz[:, 1], xyz[:, 0]), 2 * np.pi) \
            / np.pi * 180

        source_position = np.concatenate(
            (azimuth[..., np.newaxis],
             elevation[..., np.newaxis],
             radius[..., np.newaxis]), axis=-1)

    # get number of sources from data
    numSources = data.shape[1]

    # create empty SOFA object
    if mode == "HRTF":
        convention = "SimpleFreeFieldHRTF" if numSources == 2 else "GeneralTF"
    else:
        convention = "SimpleFreeFieldHRIR" if numSources == 2 else "GeneralFIR"

    sofa = sf.Sofa(convention)

    # write meta data
    sofa.GLOBAL_ApplicationName = 'Mesh2HRTF'
    sofa.GLOBAL_ApplicationVersion = Mesh2HRTF_version
    sofa.GLOBAL_History = "numerically simulated data"

    # Source and receiver data
    sofa.SourcePosition = source_position
    sofa.SourcePosition_Units = "degree, degree, meter"
    sofa.SourcePosition_Type = "spherical"

    sofa.ReceiverPosition = receiver_position
    sofa.ReceiverPosition_Units = "meter"
    sofa.ReceiverPosition_Type = "cartesian"

    # HRTF/HRIR data
    if mode == "HRTF":
        sofa.Data_Real = np.real(data)
        sofa.Data_Imag = np.imag(data)
        sofa.N = np.array(frequencies).flatten()
    else:
        sofa.Data_IR = data
        sofa.Data_SamplingRate = sampling_rate
        sofa.Data_Delay = np.zeros((1, data.shape[1]))

    return sofa


def _output_to_hrtf_load(foldername, filename, numFrequencies):
    """
    Load results of the BEM calculation.

    Parameters
    ----------
    foldername : string
        The folder from which the data is loaded. The data to be read is
        located in the folder be.out inside NumCalc/source_*
    filename : string
        The kind of data that is loaded

        pBoundary
            The sound pressure on the object mesh
        vBoundary
            The sound velocity on the object mesh
        pEvalGrid
            The sound pressure on the evaluation grid
        vEvalGrid
            The sound velocity on the evaluation grid
    numFrequencies : int
        the number of simulated frequencies

    Returns
    -------
    data : numpy array
        Pressure or abs velocity values of shape (numFrequencies, numEntries)
    """

    # ---------------------check number of header and data lines---------------
    current_file = os.path.join(foldername, 'be.1', filename)
    numDatalines = None
    with open(current_file) as file:
        line = csv.reader(file, delimiter=' ', skipinitialspace=True)
        for idx, li in enumerate(line):
            # read number of data points and head lines
            if len(li) == 2 and not li[0].startswith("Mesh"):
                numDatalines = int(li[1])

            # read starting index
            elif numDatalines and len(li) > 2:
                start_index = int(li[0])
                break

    # ------------------------------load data----------------------------------
    dtype = complex if filename.startswith("p") else float
    data = np.zeros((numFrequencies, numDatalines), dtype=dtype)

    for ii in range(numFrequencies):
        tmpData = []
        current_file = os.path.join(foldername, 'be.%d' % (ii+1), filename)
        with open(current_file) as file:

            line = csv.reader(file, delimiter=' ', skipinitialspace=True)

            for li in line:

                # data lines have 3 ore more entries
                if len(li) < 3 or li[0].startswith("Mesh"):
                    continue

                if filename.startswith("p"):
                    tmpData.append(complex(float(li[1]), float(li[2])))
                elif filename == "vBoundary":
                    tmpData.append(np.abs(complex(float(li[1]), float(li[2]))))
                elif filename == "vEvalGrid":
                    tmpData.append(np.sqrt(
                        np.abs(complex(float(li[1]), float(li[2])))**2 +
                        np.abs(complex(float(li[3]), float(li[4])))**2 +
                        np.abs(complex(float(li[5]), float(li[6])))**2))

        data[ii, :] = tmpData if tmpData else np.nan

    return data, np.arange(start_index, numDatalines + start_index)
